File: Python/351__Android_Unlock_Patterns.py
# Using DFS Solution
class Solution(object):
    def numberOfPatterns(self, m, n):
        """
        :type m: int
        :type n: int
        :rtype: int
        """
        skip = {(1,7) : 4, (1, 3) : 2, (1, 9) : 5, (2, 8) : 5, (3, 7) : 5, (3, 9) : 6, (4, 6) : 5, (7, 9) : 8}
        self.result = 0

        def dfs(used, last):
            if len(used) >= m:
                self.result += 1
            if len(used) == n:
                return
            for j in range(1, 10):
                if j not in used:   # if j is not used
                    # Sort the vertices of the edge to search in skip
                    edge = (min(last, j), max(last, j))
                    if edge not in skip or skip[edge] in used:
                        dfs(used + [j], j)

        for i in range(1, 10):
            dfs([i], i)
        return self.result
# Can be optimized by only start at 1, 2, and 5.
# result = dfs([1], 1) * 4 + dfs([2], 2) * 4 + dfs([5], 5)

File: Java/351__Android_Unlock_Patterns.java
class Solution {

    int count = 0;

    public int numberOfPatterns(int m, int n) {
        boolean[] visited = new boolean[10];
        for (int start = 1; start <= 2; start++) {
            int c = count;
            visited[start] = true;
            dfs(start, visited, 1, m, n);
            visited[start] = false;
            count = c + (count - c) * 4;
        }
        visited[5] = true;
        dfs(5, visited, 1, m, n);
        visited[5] = false;
        return count;
    }

    private void dfs(int key, boolean[] visited, int len, int m, int n){
        if (len >= m && len <= n)
            count++;
        for (int nei = 1; nei <= 9; nei++) {
            if (len == n) break;
            if (key == nei || visited[nei]) continue;
            int d = nei + key;
            if (d % 2 == 0 && (Math.abs((key - 1) / 3 - (nei - 1) / 3) > 1 || Math.abs((key - 1) % 3 - (nei - 1) % 3) > 1) && !visited[d / 2])
                continue;
            visited[nei] = true;
            dfs(nei, visited, len + 1, m, n);
            visited[nei] = false;
        }
    }
}
